runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -214,21 +213,22 @@ State waitUntilFinish(
     MonitoringUtil monitor = new MonitoringUtil(projectId, dataflowOptions.getDataflowClient());
 
     long lastTimestamp = 0;
-    BackOff backoff =
-        duration.getMillis() > 0
-            ? new AttemptAndTimeBoundedExponentialBackOff(
-                MESSAGES_POLLING_ATTEMPTS,
-                MESSAGES_POLLING_INTERVAL,
-                duration.getMillis(),
-                AttemptAndTimeBoundedExponentialBackOff.ResetPolicy.ATTEMPTS,
-                nanoClock)
-            : new AttemptBoundedExponentialBackOff(
-                MESSAGES_POLLING_ATTEMPTS, MESSAGES_POLLING_INTERVAL);
+    FlexibleBackoff backoff = FlexibleBackoff.of()
+        .withMaxRetries(MESSAGES_POLLING_RETRIES)
+        .withInitialBackoff(MESSAGES_POLLING_INTERVAL);
+    if (duration.isLongerThan(Duration.ZERO)) {
+      backoff = backoff.withMaxCumulativeBackoff(duration);
+    }
+
+    // In this function, track the cumulative time since we've started backing off and use this
+    // to enforce a wall-clock limit.
+    long startNanos = nanoClock.nanoTime();
+
     State state;
     do {
       // Get the state of the job before listing messages. This ensures we always fetch job
       // messages after the job finishes to ensure we have all them.
-      state = getStateWithRetries(1, sleeper);
+      state = getStateWithRetries(0, sleeper);
```

Comment:
The 3 custom BackOff implementations we had were very casual and inconsistent about "retries" vs "attempts". This function intended for there to be exactly 1 attempt, but said 1 retry. Hence this change (and several similar).